### PR TITLE
cpu/sam3: added RTT driver

### PR DESCRIPTION
**Pull request by haukepetersen**
Found and finished this code in one of my branches...

### CODE REVIEW

File: cpu/sam3/cpu.c [hunk 1]
```diff
@@ -90,6 +95,14 @@ void cpu_init(void)
     /* wait for master clock to be ready */
     while (!(PMC->PMC_SR & PMC_SR_MCKRDY));
 
+    /* setup the SCLK: switch to external oscillator if enabled */
+#if CLOCK_SCSLK_XTAL
+    /* enable external oscillator */
+    SUPC->SUPC_CR = (SUPC_CR_KEY(SUPCKEY) | SUPC_CR_XTALSEL);
+    while (!(SUPC->SUPC_OSCSEL & SUPC_SR_OSCSEL)) {}
+
```
Comment (aabadie) [inline cpu/sam3/cpu.c]: This extra line can be removed I think

File: cpu/sam3/periph/rtt.c [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+ *
+ * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ *
+ * @}
+ */
+
+#include "cpu.h"
+#include "periph/rtt.h"
+
+#define ENABLE_DEBUG        (1)
```
Comment (aabadie) [inline cpu/sam3/periph/rtt.c]: Debug should be disabled

**Comment by aabadie** [comment]
@haukepetersen, I gave a try to this one with `tests/periph_rtt`: it seems to work. Not sure if it's a problem or not but I noticed a (nearly) 1s time shift for each `Hello` triggered. See my output below:

2017-09-13 13:43:40,729 - INFO # Initializing the RTT driver
2017-09-13 13:43:40,730 - INFO # [rtt] setting prescaler to 32768
2017-09-13 13:43:40,731 - INFO # Setting initial alarm
2017-09-13 13:43:40,733 - INFO # [rtt] set new alarm to trigger at 10
2017-09-13 13:43:40,735 - INFO # Done setting up the RTT, wait for many Hellos
2017-09-13 13:43:51,954 - INFO # [rtt] ISR: state is 0x00000003
2017-09-13 13:43:51,958 - INFO # [rtt] set new alarm to trigger at 20
2017-09-13 **13:43:51,959** - INFO # Hello
2017-09-13 13:44:02,875 - INFO # [rtt] ISR: state is 0x00000003
2017-09-13 13:44:02,877 - INFO # [rtt] set new alarm to trigger at 30
2017-09-13 **13:44:02,878** - INFO # Hello
2017-09-13 13:44:13,792 - INFO # [rtt] ISR: state is 0x00000003
2017-09-13 13:44:13,795 - INFO # [rtt] set new alarm to trigger at 40
2017-09-13 **13:44:13,795** - INFO # Hello
2017-09-13 13:44:24,712 - INFO # [rtt] ISR: state is 0x00000003
2017-09-13 13:44:24,715 - INFO # [rtt] set new alarm to trigger at 50
2017-09-13 **13:44:24,715** - INFO # Hello

**Comment by gebart** [comment]
@aabadie check the reference manual for your CPU, it could be the same problem as for kinetis where the alarm is triggered when the counter equals the alarm time and the counter increments. See #7476, #6331

**Comment by aabadie** [comment]
Thanks for pointing me to this @gebart. It indeed makes more sense to me what behavior is expected in #6331. And I'm not the only one apparently.

**Comment by aabadie** [comment]
@haukepetersen, any chance that you addressed the comments here before feature freeze ?
We also merge #7476 I think.

**Comment by haukepetersen** [comment]
Will try. Tried to debug the behavior further last week, but so far I am completely in the blank why the RTT is to imprecise on the arduino-due. Will take more debugging...

File: cpu/sam3/periph/rtt.c [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+#include "debug.h"
+
+/* guard file to prevent file being build if no RTT was configured */
+#if RTT_NUMOF
```
Comment (smlng) [inline cpu/sam3/periph/rtt.c]: is this still needed?

Comment (haukepetersen) [reply cpu/sam3/periph/rtt.c]: nope, its not. Will remove once I find time to actually get this driver working correctly...

**Comment by smlng** [comment]
looks like this still needs some work and testing, postpone to 2018.04.

**Comment by PeterKietzmann** [comment]
@haukepetersen ping

**Comment by haukepetersen** [comment]
I am aware of this PR, though its not very high up my prio list...

**Comment by haukepetersen** [comment]
Finally found a minute: this PR should be all good now:

- code is cleaned up
- clock init is fixed
- tests/periph_rtt runs as expected

BUT note: per default, the slow clock driving the RTT is clocked by an internal `22 to 42 kHz Low Power RC Oscillator` (-> see datasheet sec. 16.2). So as observed above, the timings are quite off w.r.t. to the 5s interval the examples should run at.

It turns out (at least my google search), that many arduino-due boards do not have the external 32khz oscillator equipped, which would be needed for better timings. So I left the default configuaration to use the internal RC-oscillator. Simply compile with `CFLGAS=-DCLOCK_SCLK_XTAL=1` to switch to the external osciallator (if your board has it equipped...).

**Comment by haukepetersen** [comment]
got hold of a due board with the external oscillator equipped: 
- the init function now waits for the prescaler to be applied
- fixed a 1-off error when setting the alarm value

Now the output of the `periph_rtt` test looks like expected (building with `CFLAGS=-DCLOCK_SCLK_XTAL`:
```
2020-03-11 11:13:15,401 # main(): This is RIOT! (Version: 2020.04-devel-1110-g4113a-add_sam3_rtt)
2020-03-11 11:13:15,401 # 
2020-03-11 11:13:15,402 # RIOT RTT low-level driver test
2020-03-11 11:13:15,406 # This test will display 'Hello' every 5 seconds
2020-03-11 11:13:15,406 # 
2020-03-11 11:13:15,410 # Initializing the RTT driver
2020-03-11 11:13:15,410 # RTT now: 0
2020-03-11 11:13:15,414 # Setting initial alarm to now + 5 s (5)
2020-03-11 11:13:15,418 # Done setting up the RTT, wait for many Hellos
2020-03-11 11:13:20,410 # Hello
2020-03-11 11:13:25,408 # Hello
2020-03-11 11:13:30,409 # Hello
2020-03-11 11:13:35,410 # Hello
2020-03-11 11:13:40,411 # Hello
2020-03-11 11:13:45,409 # Hello
2020-03-11 11:13:50,410 # Hello
```

File: boards/common/arduino-due/include/periph_conf.h [hunk 4]
```diff
@@ -62,6 +72,13 @@ static const timer_conf_t timer_config[] = {
 #define TIMER_NUMOF         ARRAY_SIZE(timer_config)
 /** @} */
 
+/**
+ * @name    RTT configuration
+ * @{
+ */
+#define RTT_FREQUENCY       (1U)        /* 1Hz */
```
Comment (kaspar030) [inline boards/common/arduino-due/include/periph_conf.h]: Could you default this to 1024 or 32768?

**Comment by kaspar030** [comment]
Tested using `tests/periph_rtt`, works fine! I'm not entirely happy that this important fix is optional and hidden behind a global define, but at least now there's a way to use it somehow.
I guess there's no easy way to check if the XTAL is present?

File: boards/common/arduino-due/include/periph_conf.h (hunk 4 above)
Comment (kaspar030) [reply boards/common/arduino-due/include/periph_conf.h]: scratch that, need to solve this globally.

**Comment by haukepetersen** [comment]
> I guess there's no easy way to check if the XTAL is present?

nope, not that I know of.

**Comment by haukepetersen** [comment]
should I squash?

**Comment by haukepetersen** [comment]
squashed.

**Comment by aabadie** [comment]
@fjmolinas can you test this PR ? There should be an arduino-due on my desk.

**Pull request merged**